### PR TITLE
append warning for stale data (#46)

### PR DESCRIPTION
**Pull request by ethanlabelle**
This is a small PR to add a warning to all RPC responses when Clio has not published a ledger for 60 seconds. 
```
Manual Testing Strategy (HTTP requests):
- Start rippled, wait for its ledger to sync
- Start clio
- Send clio http requests: 
    - i.e.  curl localhost:51233 -d {"method": "server_info"}' | jq 
- Assert no warnings
- Stop rippled process
- Wait one minute
- Send clio http requests
- Assert warnings
- Turn on rippled, wait for sync
- Send requests to clio
- Verify warning disappears when clio starts publishing ledgers
```

The testing procedure for websocket connections is nearly identical. Instead of using curl to send http requests, create a ws connection using a tool such as wscat.

### CODE REVIEW

File: src/etl/ReportingETL.h [hunk 1]
```diff
@@ -130,19 +130,12 @@ class ReportingETL
...
-        std::unique_lock<std::mutex> lck(publishTimeMtx_);
-        return lastPublish_;
-    }
+    mutable std::shared_mutex publishTimeMtx_;
 
     void
     setLastPublish()
     {
-        std::unique_lock<std::mutex> lck(publishTimeMtx_);
+        std::unique_lock<std::shared_mutex> lck(publishTimeMtx_);
```
Comment (natenichols) [inline src/etl/ReportingETL.h]: nit: lets remove this template. Let the compiler infer the type of the mutex.

```cpp
std::unique_lock lck(publishTimeMtx_);
```

File: src/etl/ReportingETL.h [hunk 2]
```diff
@@ -329,6 +322,13 @@ class ReportingETL
 
         return result;
     }
+
+    std::chrono::time_point<std::chrono::system_clock>
+    getLastPublish() const
+    {
+        std::shared_lock<std::shared_mutex> lck(publishTimeMtx_);
```
Comment (natenichols) [inline src/etl/ReportingETL.h]: Same here.

File: src/webserver/HttpBase.h [hunk 3]
```diff
@@ -405,8 +405,27 @@ handle_request(
             responseStr = boost::json::serialize(response);
         }
 
+        auto warningFlag = false;
+        result["warning"] = boost::json::array{};
```
Comment (natenichols) [inline src/webserver/HttpBase.h]: You could try something like:
```cpp
boost::json::array& warning = result["warning"].as_array();
....
warning.emplace_back("Too many requests");
```

These are pretty much functionally equivalent, so feel free to leave it as is, too.

File: src/webserver/HttpBase.h [hunk 4]
```diff
@@ -405,8 +405,27 @@ handle_request(
...
+                context->etl->getLastPublish())
+                .count();
+        if (lastPublishAgeMinutes >= 1)
+        {
+            result["warning"].as_array().emplace_back(
+                "This server may be out of date");
+            warningFlag = true;
+        }
+        // reserialize only if a warning was appended.
+        if (warningFlag)
```
Comment (natenichols) [inline src/webserver/HttpBase.h:422]: Good catch on this bug.

File: src/webserver/WsBase.h [hunk 5]
```diff
@@ -320,9 +321,28 @@ class WsSession : public WsBase,
...
+        result["warning"] = boost::json::array{};
         std::string responseStr = boost::json::serialize(response);
-        dosGuard_.add(*ip, responseStr.size());
+        auto warningFlag = false;
+        if (!dosGuard_.add(*ip, responseStr.size()))
+        {
+            result["warning"].as_array().emplace_back("Too many requests");
+            warningFlag = true;
+        }
+        auto lastPublishAgeMinutes =
```
Comment (natenichols) [inline src/webserver/WsBase.h]: We may want to consider putting this as a member in `ReportingETL` so we don't have to do this `duration_cast` everywhere we want this info from the etl. 

```cpp
std::uint32_t 
ReportingETL::lastPublishAgeSeconds()
{
    return std::chrono::duration_cast<std::chrono::minutes>(
                 std::chrono::system_clock::now() - getLastPublish())
                 .count();
}
```

If you chose to go this route, also call this in `getInfo()`

**Review by natenichols — APPROVED** [review]
Nice work! None of my comments should hold up approval. 

I'm pretty indifferent towards my comments. If you like the suggestions, feel free to implement them. Else, leave a comment why you disagree.

File: src/webserver/HttpBase.h (hunk 3 above)
Comment (ethanlabelle) [reply src/webserver/HttpBase.h]: I chose to take a slightly different approach that I believe is very clear.
```
boost::json::array warnings; // creates empty json array
...
warnings.emplace_back("Too many requests");
...
result["warning"] = warnings;
```
This way, the warnings array is not connected to the response json object until after all warnings, if any, have been added. Just need to make sure that result is a reference to response["result"] 😉

**Pull request merged**